fix #8: send auth token as bearer token

kincaidoneil opened the pull request:
- Send auth token as bearer token to fix #8 
- Breaking change to bring into compliance with [HTTP bearer auth](https://swagger.io/docs/specification/authentication/bearer-authentication/)

kincaidoneil committed:
fix #8: send auth token as bearer token

emschwartz reviewed (approved):
Note this is a breaking change but it's an important one to make because this plugin does not use the Authorization header in the correct way

matdehaast commented:
@emschwartz this would be a a breaking change for live connectors. @sharafian

sharafian reviewed (changes requested):
this should be behind a flag in order to make it non breaking. I don't see a compelling reason to make this change breakingly

sharafian commented:
we could just have `secretBearerToken` instead of `secretToken` and advise everyone use that. Then it doesn't break current usage at all.

emschwartz commented:
I'm not sure how much value it adds to make this change in a non-breaking way. You could change the config option name but that still doesn't change the fact that an older version would reject the auth details.

I would just bump the major version so no one updates unintentionally.

Part of the motivation for this is that we want to peer `ilp-connector` with the other implementations but don't want to add an incorrect auth method just to be compatible with how this plugin handles auth. Also, not including the `Bearer` part makes it harder to distinguish different auth methods we may want to support.

sharafian commented:
Bumping the major version means that we'll need to switch to the new breaking version to stay up to date on fixes and patches. Breakingly changing the communication between all of our connectors is something we don't really want to do in production, especially because there's no upside.

> You could change the config option name but that still doesn't change the fact that an older version would reject the auth details.

This is true no matter what. I'd argue it's more confusing if peering `secretToken` to `secretToken` doesn't work (only when versions are mismatched) vs. peering `secretToken` to `secretBearerToken` not working.

kincaidoneil committed:
fix: new config options for bearer tokens

kincaidoneil commented:
Updated with new configuration options for bearer tokens and updated the README with these configs. Also updated the tests for each of the 4 different auth configurations.

sharafian commented on src/index.ts:
```diff
@@ -87,13 +101,15 @@ class PluginHttp extends EventEmitter {
+      staticToken: incoming.secretToken || incoming.staticToken
     })
+
+    this._useBearerToken = !outgoing.secret || !outgoing.secretToken
```
this will send `Bearer ` if we've specified `secretToken` but not `secret`. It should only send `Bearer ` if staticToken is set.

Maybe instead you could do

```js
  staticToken: incoming.staticToken ? `Bearer ${incoming.staticToken}` : incoming.secretToken
```

and then you never have to add on the `Bearer ` at request time

kincaidoneil committed:
fix: bug using bearer with old config

kincaidoneil committed:
fix: vulnerabilities

kincaidoneil reviewed (commented):
Nice catch! Fixed.

The intended behavior is to append `Bearer ` if none of the old config options were set (`secretToken` nor `secret`). I believe JWTs are also supposed to have bearer prepended to the generated token, which is why it's done per request (otherwise I think it'd require changes to the auth classes)

kincaidoneil commented:
GitHub wasn't showing the recent commits to the base branch in the PR; closing and re-opening seemed to fix it  🤔

matdehaast commented:
> GitHub wasn't showing the recent commits to the base branch in the PR; closing and re-opening seemed to fix it 🤔

Github had outages 12hrs after this PR. Wonder if the issues actually started occurring earlier than reported and hence no history being propagated

sharafian commented on src/index.ts:
```diff
@@ -161,12 +177,18 @@ class PluginHttp extends EventEmitter {
-  _verifyToken (token: string): Promise<boolean> {
+  _verifyAuth (authHeader: string): Promise<boolean> {
+    const token = authHeader.startsWith('Bearer ')
+      ? authHeader.replace('Bearer ', '')
```
nit `authHeader.substring('Bearer '.length)` may be a little faster because it doesn't require matching

kincaidoneil committed:
fix: faster string replacement
